### PR TITLE
introduce INF-RT

### PR DESCRIPTION
**Pull request by Superjomn**
### PR types
New features

### PR changes
Others

### Describe
Add inf-rt code.

INF-RT(short for Inference Runtime) is a new architecture targeting the next generation inference engine.

These codes are moved from https://github.com/PaddlePaddle/CINN/tree/develop/infrt, it has been developed since the 10-2020, once merged, the code will be removed from there.

### CODE REVIEW

File: paddle/infrt/host_context/core_runtime_test.cc [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace infrt {
+namespace host_context {
+
+int add(int a, int b) { return a + b; }
+int sub(int a, int b) { return a - b; }
+
+TEST(CoreRuntime, basic) {
+  KernelRegistry registry;
+  registry.AddKernel("cinn.test.addi32", CINN_KERNEL(add));
```
Comment (phlrain) [inline paddle/infrt/host_context/core_runtime_test.cc]: 有没有处理tensor的kernel的单测的例子

Comment (Superjomn) [reply paddle/infrt/host_context/core_runtime_test.cc]: 这是一个 int32 + int32 的 scalar 的kernel

其kernel 实现：

int32_t add(int32_t, int32_t)

Comment (Superjomn) [reply paddle/infrt/host_context/core_runtime_test.cc]: 这里手动调用了该 kernel，做了验证

https://github.com/PaddlePaddle/Paddle/pull/37669/files/a16035a9418d223c24e7cae3b7aa99fdeafa3511#diff-b05d0bc33e47163195ee4a8b20cc395679ea9baa94d98dd2623b62bf0ea7a51cR42

File: paddle/infrt/host_context/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+cc_test_tiny(test_infrt_op_executable SRCS op_executable_test.cc DEPS infrt ${MLIR_IR_LIBS})
+cc_test_tiny(test_infrt_core_runtime SRCS core_runtime_test.cc DEPS infrt ${MLIR_IR_LIBS})
+cc_test_tiny(test_infrt_mlir_to_runtime_translate SRCS mlir_to_runtime_translate_test.cc DEPS infrt ${MLIR_IR_LIBS})
+
+cinn_exec_check(test_infrt_mlir_exec_on_basic mlir_tests/basic.mlir)
+cinn_exec_check(test_infrt_mlir_exec_on_shape mlir_tests/shape.mlir)
+cinn_exec_check(test_infrt_mlir_exec_on_dense_tensor mlir_tests/dense_tensor.mlir)
+
+add_executable(cinn-exec mlir_exec.cc)
+target_link_libraries(cinn-exec infrt ${MLIR_IR_LIBS})
```
Comment (winter-wang) [inline paddle/infrt/host_context/CMakeLists.txt]: 这个cinn-exec是不是也改为  infrt-exec比较好？

**Review by Shixiaowei02 — DISMISSED** [review]

File: paddle/infrt/host_context/CMakeLists.txt (hunk 2 above)
Comment (Superjomn) [reply paddle/infrt/host_context/CMakeLists.txt]: done

Replaced globally.

**Review by Shixiaowei02 — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM for const_cast

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**